[RFC] Make archival asynchronous

kevans91 opened the pull request:
The prime benefit being sought here is for large archives to not
clog up the rendering process and cause unsightly proxy timeouts.
As a secondary benefit, archive-in-progress is moved out of the
way into a /tmp file so that new archival requests for the same
commit will not get fulfilled based on an archive that isn't yet
finished.

Fixes #11265

This decidedly needs, at a minimum:
- [] Some tests to go with it
- [] Perhaps an actual "loading" icon... right now it slaps a loading class on the download icon, sending that sucker spinning.

I've deployed this to my production git instance; the benefit can be seen here: https://git.kevans.dev/kevans/freebsd/branches/ -> any of these branches will take probably 2/3 minutes to finish archiving, so the user experience is slightly improved by making this async. There's a cron job on that instance that will wipe out any generated archives every 15 minutes, so it should be easy to reproduce.

Ultimately, all archival requests get pushed into a queue. The queue handler will asynchronously process all of the requests within it and periodically purge the local archiveInProgress slice. This probably didn't need to be a queue in the end, as we ended up not using the queue very effectively; a later refactor may just move the queue handler into ArchiveRepository().

lafriks commented:
why deleting them every 15 minutes? imho that is way too fast

kevans91 commented:
> why deleting them every 15 minutes? imho that is way too fast

Ah, this is just for testing, since I only have a handful of branches on this repo for easily checking the behavior when a branch takes too long to archive. I can relax this, but the main point was just to invalidate them quickly enough that I don't need to manually purge. =)

guillep2k commented:
Hmm... I've just browsed the code a little, but I think this will affect robots and other automated calls that require the link to resolve itself in the same request. Maybe we should keep the current endpoint for that purpose and use a different URL for the UI.

kevans91 commented:
> Hmm... I've just browsed the code a little, but I think this will affect robots and other automated calls that require the link to resolve itself in the same request. Maybe we should keep the current endpoint for that purpose and use a different URL for the UI.

I naively hadn't considered the possibility, but as I had re-assessed what I've done more I did come to realize that the current endpoint didn't really need to change like it did. This seems like a compelling enough reason to fully revert it to being synchronous (enqueue then wait for completion) -- they'd still reap the benefit of never seeing an incomplete archive, even if such a thing would likely timeout the first time they attempted to fetch it.

guillep2k commented:
There's always the option -to a point- of sending a [periodic redirect](https://stackoverflow.com/a/36041063/225979). 😏

guillep2k commented on routers/repo/repo.go:
```diff
@@ -458,80 +455,79 @@ func RedirectDownload(ctx *context.Context) {
+	complete := aReq.IsComplete()
+	ctx.JSON(200, map[string]interface{}{
+		"archiving": !complete,
+		"complete":  complete,
```
We could drop one flag.

guillep2k commented on routers/routes/routes.go:
```diff
@@ -878,7 +878,11 @@ func RegisterRoutes(m *macaron.Macaron) {
 
-		m.Get("/archive/*", repo.MustBeNotEmpty, reqRepoCodeReader, repo.Download)
+		m.Group("/archive", func() {
+			m.Post("/status/*", repo.DownloadStatus)
```
Can this generate a name conflict? (e.g. a user named `status`)

kevans91 replied:
Shouldn't- this is a little further down in the chain: `/:username/:reponame/archive/status/*`

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,309 @@
+			ctx.ServerError("GetTagCommit", err)
+			return nil
+		}
+	} else if len(r.refName) >= 4 && len(r.refName) <= 40 {
```
We should filter values like `heads/pulls/345`. There's a Regexp somewhere that checks if a string is a valid commit sha1 (`^[0-9a-f]{4,40}$`).

kevans91 replied:
I went ahead and went with {4,64} to remove the need to change it if one manages to use gitea with a sha256-enabled repo, but will reduce that back to 40 on request...

guillep2k commented on routers/repo/repo.go:
```diff
@@ -458,80 +455,79 @@ func RedirectDownload(ctx *context.Context) {
+	aReq := archiver_service.DeriveRequestFrom(ctx, uri)
+
+	downloadName := ctx.Repo.Repository.Name + "-" + aReq.GetArchiveName()
+	if !aReq.IsComplete() {
+		aReq = archiver_service.ArchiveRepository(aReq)
+		archiver_service.LockQueue()
+		for !aReq.IsComplete() {
+			archiver_service.WaitForCompletion()
 		}
+		archiver_service.UnlockQueue()
 	}
```
I think this block can be moved to the archiver service, so the lock semantics remain opaque.

guillep2k commented on routers/repo/repo.go:
```diff
@@ -458,80 +455,79 @@ func RedirectDownload(ctx *context.Context) {
+		for {
+			if aReq.IsComplete() || time.Now().After(timeout) {
+				break
+			}
```
This spin loop is taxing for the CPU and a probable door for a DOS attack. The "golang way" should be to use a `chan`, but we could make do with a thread sleep:
```
time.Sleep(250 * time.Millisecond)
```

kevans91 replied:
To be honest, what we really want here is a `sync.Cond.TimedWait` -- perhaps something like this: https://github.com/golang/go/issues/9578#issuecomment-69700942... this would entail the extra goroutine to Cond.Wait() and close the channel if it returns, but is that a problem?

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,309 @@
+	// and it is not marked complete.
+	archiveMutex.Lock()
+	if rExisting := getArchiveRequest(request.repo, request.commit, request.archiveType); rExisting != nil {
+		archiveMutex.Unlock()
```
I think we probably can just `defer` the unlock.

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,309 @@
+		// Wake up all other goroutines that may be waiting on a request to
+		// complete.  They should all wake up, see if that particular request
+		// is complete, then return to waiting if it is not.
+		archiveCond.Broadcast()
```
This is what channels (`chan`) are for 😉

kevans91 replied:
... but now that I read this particular comment, the solution makes more sense and doesn't involve Cond at all. :-)

kevans91 commented:
Here's a checkpoint with most of your comments, I think, addressed... along with an attempt at fixing the test after re-working waiting for completion.

kevans91 commented:
I applied a little bit of critical thinking and realized that of course the /archive/status endpoint could induce a naming headache... I then applied even more and decided it was worthless anyways, so it's been removed.

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,306 @@
+// It returns whether the archive was actually completed, as the channel could
+// have also been closed due to an error.
+func (aReq *ArchiveRequest) WaitForCompletion() bool {
+	<-aReq.cchan
+	return aReq.IsComplete()
```
```suggestion
	import "code.gitea.io/gitea/modules/graceful"

	graceful.GetManager().RunWithShutdownContext(func(ctx context.Context) {
		select {
		case <-aReq.cchan:
			return aReq.IsComplete()
		case <-ctx.Done():
			return false
		}
	}
```
We should not block Gitea from shutting down. 😄 
Note: the `import` goes at the top of the file, of course.

kevans91 replied:
I think this (and the timeout comment below) is addressed, but I'll leave it unresolved for now in case there's follow-up -- I'm not 100% certain about the approach I ended up taking, but I think it's OK since we'll still terminate and promptly close the locally-constructed wait channel.

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,306 @@
+		return aReq.IsComplete(), true
+	case <-aReq.cchan:
+		return aReq.IsComplete(), false
+	}
```
Even if we expect `dur` to be short, we should run this in a "with shutdown" context as well.

guillep2k commented on services/archiver/archiver.go:
```diff
@@ -0,0 +1,306 @@
+		if idx != lastidx {
+			archiveInProgress[idx] = archiveInProgress[lastidx]
+		}
+		archiveInProgress = archiveInProgress[:lastidx]
```
Instead of swapping, you could just:
```
archiveInProgress = append(archiveInProgress[:idx], archiveInProgress[idx+1:]...);
```

https://play.golang.org/p/w1-d3aaCola

kevans91 replied:
So, I wasn't entirely sure here if I should be concerned about potential performance implications from slicing twice around the element, vs. the trivial swap to the end and cut it off.

guillep2k replied:
Slices in go are not arrays; this kind of manipulation is resolved internally using chunks and pointers (that's why they've got a name different from `array`). Super-performant.

guillep2k replied:
But we're also expecting a very short list of elements here.

kevans91 replied:
Right. ideally this queue should drain pretty quickly, but even for bots crawling somewhat large-ish installations... that's still pretty low order of magnitude. Readability is a little more important here, so I've switched it up.

kevans91 commented:
Thanks for the feedback- I'm hoping to drum up some time this week to incorporate it. =)

kevans91 commented:
Following discussion with @zeripath on Discord (Thanks!), I've simplified the logic that ensures we aren't blocking shutdown to remove the new channels and extra goroutines.

zeripath commented:
Hi @kevans91 how's this PR progressing?

kevans91 commented:
> Hi @kevans91 how's this PR progressing?

As far as I'm aware, I've addressed all of the feedback and am awaiting further review- I'd left one unresolved because I wasn't sure if the solution was satisfactory or not. We discussed it a bit, now I'm just unsure if I need to change those to use a different context (re: #12560

lafriks commented:
Please resolve conflicts

kevans91 commented:
Heh, a minute after I pushed a version with conflicts resolved, another set of conflicts popped up. =D

I think I've reconciled the differences correctly; WaitForCompletion/TimedWaitForCompletion now use Done() channels on the request context, and the service itself passes shutdown context to CreateArchive().

lafriks commented:
:rocket:
